makeStyles spec (#16631)
microsoft/fluentui: specs/makeStyles.md

### v0 approach

`@fluentui/react-northstar` (aka v0 aka Stardust) uses CSS-in-JS. [FelaJS](https://github.com/robinweser/fela) CSS-in-JS library is used.

#### Fela uses atomic CSS classes

- one CSS property = one CSS class
- all selectors have one class specificity
- the same classes are reused anywhere in the application = smaller DOM

#### CSS classes are added to DOM when used for the first time

- only the classes that are used are added to DOM
- that means those are added in non-deterministic order

#### Two CSS classes setting the same property cannot be applied to the same element

If there are two classes setting the same CSS property (`.r {color: red}` and `.b {color: blue}` those can never be applied to the same element.
As both classes have the same one class specificity, the rule defined later in the DOM will win. As the insertion order is non-deterministic, the resulting style would be non-deterministic as well.
As a consequence, there is a couple of restrictions:

- Shorthand css properties (`margin`) must be expanded before merging (to `margin-top`, `margin-right`, `margin-bottom`, `margin-left`).

> - To apply any overrides, it is not possible to just concatenate classes. Style objects must be always deep merged and then the classes can be computed.

Reviewer: I think that we are missing an explanation why deep merge is required (because of nested and pseudo selectors)

  If that is a non-FUI component, parent must pass the overrides as a list of classnames. If the child is a FUI component, the overrides must me passes as `styles` object so that the child can merge the styles correctly.
  This can be complicated as there are scenarios where the parent does not know what child is will render (especially when the styles are not applied by the direct child but somewhere else down the tree).
- When overrides are passed from parent to child, the whole styles must be re-evaluated in runtime every time. Child can cache evaluated styles and classnames based on (limited) set of input props and state. And FUI components do that. But once style overrides are passed to the component the overrides and component styles object must be deep merged.

#### Application overrides are part of the theme object

All application-owned style overrides are part of the Theme object. Most of those are just one-off overrides.
This results in a huge object (several files and thousands of lines per component) which is not tree-shakeable.

> There is also unnecessary and confusing indirection when defining overrides and no support for dead code elimination.

Reviewer: Should we add an example of this?

> There is also unnecessary and confusing indirection when defining overrides and no support for dead code elimination.

Reviewer 2: I would say that the indirection is due to the styles not living close enough to the component using them

#### No special prop for passing overrides to children

For handling multiple levels of overrides, instead of using a special prop with all its problems (see above), `useCSS` just returns a class which can be passed to a child using `className`. If there is another `useCSS` call in the child, it takes the `className` from parent.

1. `useCSS` call in parent generates a classname, and adds mapping from the classname to a style object to a global dictionary.
2. Parent takes the class returned by the `useCSS` call and passes it to its child.
3. The child calls `useCSS` to apply its own overrides and also passes it the class from parent.
4. `useCSS` finds the style object in the dictionary, merges the original style object with the child's style overrides and returns the new classname. Instead of performing an expensive object deep merge, it just concatenates the two stringified style objects and depends on the-latest-wins CSS rule.

![](makeStyles-useCss.png)

#### Limitations

This approach looks good for style overrides, we are not sure this approach works the best for core component styles. Without that, we would have two different styling approaches in place.

> Although the object string concatenation is performant, it results in a lot of CSS code.

Reviewer: ```suggestion
Although the string concatenation is performant, it results in a lot of CSS code.
```

?

There are two main functions in this approach `makeStyles` and `useStyles`.
`makeStyles` is a factory which takes style object as input and returns a `useStyles` hook. `makeStyles` does not accept any theme or React context and should be called in a module scope, not in a scope of a React component.
When rendering a component, `useStyles` hook is called passing a theme to it. The `useStyles` function injects the styles into DOM (if not done yet) and maps props to classnames.

#### Multiple classes are applied to a single element

As in static CSS approach, multiple semantic classes are applied to an element.
For example `<Button primary disabled>` renders as `<button class="primary disabled">`.
To style the component, combined classes are used in CSS as well:

```css
.primary {
}
.disabled {
}
.primary.disabled {
}
```

This is causing problems with overrides - assuming it is guaranteed that override classes are written bellow the component styles in the DOM, for most cases in this example, it is enough to have 1 class specificity for overrides to work correctly. The only case where the 1 class override does not work is for CSS properties in `.disabled.button`.

> Consumer does not know what specificity must be used for the overrides, this can also change between library versions.

Reviewer 2: Perhaps that 2 different engineers can break each other's styling ? i.e. the specificity wars

#### Stylesheet is generated build time

There is a build time step which generates stylesheet.
In runtime only the classes are applied based on props and state.

#### Styles precedence is set by module evaluation order

The order of `makeStyles` calls defines the DOM insertion order. When `makeStyles` is called, it does NOT insert the CSS to DOM, but assigns a counter-based id to the associated styles. When `useStyles` returned by the `makeStyles` is called for the first time, it inserts the CSS to DOM in the order defined by the `makeStyles` call.

> For that to work as expected, you cannot export `makeStyles` to share styles in multiple different modules - that would lead to non-deterministic insertion order. The only way to share styles is to export the style object and have a dedicated `makeStyles` calls in all the modules which reuse the style object.

Reviewer: Should we also add notes about async API (React.Suspense)?

https://codesandbox.io/s/kind-davinci-xvwxg

#### Inspired by Material UI

The approach is inspired by Material UI
There is a build step which generates a stylesheet for every `makeStyles` call.

## Converged solution

Can we use v8 `makeStyles` (build time style evaluation) but let JS handle the "specificity"?

The expensive part in v0 styling approach is the style evaluation - processing styles, expanding CSS shorthands, vendor prefixing, RTL, generating classnames. But we can do most of this build time.

### Basic algorithm

As an input, there is a style object:

```js
source = {
  color: 'red',
  margin: '0 10px 0 0',
  ':hover': {
    background: 'green',
  },
};
```

In build time, this object can be processed (CSS expand, RTL...) and atomic CSS generated for each CSS property:

```js

> buildOutput = {

Reviewer: ```suggestion
// ⚠️ Not a real build output
buildOutput = {
```

### Applying theme to styles

No matter what theme is used, the component styles are always the same. The only way how to change the component appearance is through tokens which can be used in style values.
The `styles` passed to `makeStyles([[matcher, styles]])` call can be either an object or a function. The function is called with all theme tokens.
Those tokens are resolved to CSS variable usages in build time. `ThemeProvider` component is responsible for setting the CSS variables in DOM and changing them when theme changes. When theme is switched, only the variables are changed, all styles remain the same.

::Risk:: What if we will not be able to describe the theme changes using just different values?
::Risk:: A theme consist of hundreds or even thousands of tokens. We need to verify browsers can handle this amount of CSS variables.

#### IE11 compatibility

IE11 does not support CSS variables. If we decide to support IE11, we will split the styles to the items which use tokens and those which do not. The ones which do not use tokens will be processed build time, for the ones with tokens, we will fall back to runtime evaluation.

### Hash-based classnames

> We will use a hash (MurmurHash2) of the CSS property name and the value to generate a classname. That results in deterministic (=stable) classnames. With that we can support server side rendering.

Reviewer: ```suggestion
We will use a hash (MurmurHash2) of the CSS property name and the value to generate a classname. That results in deterministic (=stable) classnames. With that we can support server side rendering without issues with hydration (#3692).
```

### Passing overrides to children using `className`

> To avoid the ambivalence of using `className` vs `styles` props when passing overrides to children based on their type, we will use the dictionary sideeffect described in `useCSS` approach above.

Reviewer: ```suggestion
To avoid the ambivalence of using `className` vs `styles` props when passing overrides to children based on their type, we will use the dictionary side effect described in `useCSS` approach above.
```

Parent component passes its overrides to a child as a list of classes in `className` prop. The child, takes the input classes and references the global dictionary to get back the necessary CSS attribute names to be able to perform the merge.

#### Alternative 1: encode the CSS property name to the hash

Instead of using the global dictionary side effect, we can use part of the hash in the classname to encode the CSS property:
![](makeStyles-classname.png)

Then when merging in a child, we can merge the classes by replacing previous classnames with the same CSS property hash.
This approach makes the classnames longer and according to the test does not make the merging any faster.

#### Alternative 2: use special `styles` prop for overrides

If the global dictionary does not perform as expected, we might return back to the previous v0 approach where we use a dedicated `styles` prop to pass overrides from parent to child.

### Another libraries using similar approach

This has been inspired by [Facebook stylex talk](https://www.youtube.com/watch?v=9JZHodNR184).

> Similar approach is also used in Attlasian `compiled` v5.

Reviewer: ```suggestion
Similar approach is also used in Attlasian [`compiled`](https://github.com/atlassian-labs/compiled) v5.
```

## Sample Code

```jsx

// Button.tsx
const useButtonStyles = makeStyles([

> 	[null, { cursor: 'pointer' }], // always applied

Reviewer: ```suggestion
  [null, { cursor: 'pointer' }], // always applied
```

formatting gfix

>   [ props => props.primary, tokens => ({ backgroundColor: tokens.colors.primary.background })],
>   [ props => pros.disabled, { backgroundColor: 'gray' }]

Reviewer: ```suggestion
  [props => props.primary, tokens => ({ backgroundColor: tokens.colors.primary.background })],
  [props => pros.disabled, { backgroundColor: 'gray' }]
```

formatting fix

]);

const Button = ({className, ...props}) => {

> 	const classes = useButtonStyles(props, className);
> 	return <button className={classes} {...props} />;

Reviewer: ```suggestion
  const classes = useButtonStyles(props, className);
  return <button className={classes} {...props} />;
```

}

// ButtonUsage.tsx
const useButtonOverrides = makeStyles([
  [null, { backgroundColor: 'red' ],
]);

const ButtonUsage = () => {
  const buttonOverrides = useButtonOverrides({});

  return <Button classname={buttonOverrides}>I have always red background</Button>
}
```

> 

Reviewer: Should we modify an example to show `ax()` function usage?

https://github.com/microsoft/fluentui/pull/16411

How would that be implemented.
Example API:

```js
// The Button styles need an id to pull overrides
// from the theme:
const makeStyles({ id: 'Button', { ... } });

// The theme needs to be able to define them:
const theme = {
  styles: {
    Button: { // id?
      styles: makeStyles({ ... }) // maybe?
      }
    }
  }
}
```

### Always matching matcher

What is the best signature for a `makeStyles` call with a single style object which always matches?

```js
makeStyles([[null, { background: 'red ' }]]);
makeStyles([{ background: 'red' }]);
makeStyles({ background: 'red' });
makeOverrides({ background: 'red' });
```

### `makeStyles` per component or per slot

Styling a component with three slots, do we need to have three `makeStyles` and three `useStyles` calls or can we pass an object with all slots?

### CSS variables and shorthand expand

Having a token which defines multiple values for a CSS shorthand, how can we expand the CSS property?

```js
fourValuesPaddingToken = '1 2 3 4' // results in --four-values-padding: 1 2 3 4

padding: var(--four-values-padding); // how can this be expanded to padding-top, padding-right...
```

### Impact on bundle size

> What is the impact on bundle side having all the expanded styles in bundle?

Reviewer: ```suggestion
What is the impact on bundle side having all the expanded styles in bundle?
Can we hash property keys (`display`,`display:hover`)? 
```

# makeStyles RFC

## Background

`makeStyles` is a way how CSS is applied to core components and their variants as well as a utility used to implement style overrides for those components.

## Prior Art

This section summarises previous approaches and their pros and cons.

### v0 approach

> `@fluentui/react-northstar` (aka v0 aka Stardust) uses CSS-in-JS. [FelaJS](https://github.com/robinweser/fela) CSS-in-JS library is used.

Reviewer: ```suggestion
`@fluentui/react-northstar` (aka v0 aka Stardust) uses CSS-in-JS. [Fela](https://github.com/robinweser/fela) CSS-in-JS library is used.
```

nit: I think it's just Fela

#### Fela uses atomic CSS classes

- one CSS property = one CSS class
- all selectors have one class specificity
- the same classes are reused anywhere in the application = smaller DOM

#### CSS classes are added to DOM when used for the first time

- only the classes that are used are added to DOM
- that means those are added in non-deterministic order

#### Two CSS classes setting the same property cannot be applied to the same element

> If there are two classes setting the same CSS property (`.r {color: red}` and `.b {color: blue}` those can never be applied to the same element.

Reviewer: ```suggestion
If there are two classes setting the same CSS property (`.r {color: red}` and `.b {color: blue}`) those can never be applied to the same element.
```

As both classes have the same one class specificity, the rule defined later in the DOM will win. As the insertion order is non-deterministic, the resulting style would be non-deterministic as well.
As a consequence, there is a couple of restrictions:

- Shorthand css properties (`margin`) must be expanded before merging (to `margin-top`, `margin-right`, `margin-bottom`, `margin-left`).
- To apply any overrides, it is not possible to just concatenate classes. Style object are deeply nested as they can contain styles for pseudo elements (`::before`), pseudo classes (`:hover`) or nested elements (`& .indicator`). Therefore style objects must be always deep merged and then the classes can be computed.
- Due to the previous point, when parent component is passing style overrides to a child component it needs to know what kind of component the child is.

>   If that is a non-FUI component, parent must pass the overrides as a list of classnames. If the child is a FUI component, the overrides must me passes as `styles` object so that the child can merge the styles correctly.

Reviewer 3: ```suggestion
  If that is a non-FUI component, parent must pass the overrides as a list of classnames. If the child is a FUI component, the overrides must be passed as `styles` object so that the child can merge the styles correctly.
```

>   This can be complicated as there are scenarios where the parent does not know what child is will render (especially when the styles are not applied by the direct child but somewhere else down the tree).

Reviewer 3: ```suggestion
  This can be complicated as there are scenarios where the parent does not know what child it will render (especially when the styles are not applied by the direct child but somewhere else down the tree).
```

> - When overrides are passed from parent to child, the whole styles must be re-evaluated in runtime every time. Child can cache evaluated styles and classnames based on (limited) set of input props and state. And FUI components do that. But once style overrides are passed to the component the overrides and component styles object must be deep merged.

Reviewer: ```suggestion
- When overrides are passed from parent to child, the whole styles must be re-evaluated in runtime every time. Child can cache evaluated styles and classnames based on (limited) set of input props and state. And FUI components do that. But once style overrides are passed to the component the overrides (`styles` or `variables`) and component styles object must be deep merged.
```

> - When overrides are passed from parent to child, the whole styles must be re-evaluated in runtime every time. Child can cache evaluated styles and classnames based on (limited) set of input props and state. And FUI components do that. But once style overrides are passed to the componentm the overrides (`styles` or `variables`) and component styles object must be deep merged.

Reviewer: ```suggestion
- When overrides are passed from parent to child, the whole styles must be re-evaluated in runtime every time. Child can cache evaluated styles and classnames based on (limited) set of input props and state. And FUI components do that. But once style overrides are passed to the component the overrides (`styles` or `variables`) and component styles object must be deep merged.
```

A small typo

#### Application overrides are part of the theme object

All application-owned style overrides are part of the Theme object. Most of those are just one-off overrides.
This results in a huge object (several files and thousands of lines per component) which is not tree-shakeable.
There is also unnecessary and confusing indirection when defining overrides and no support for dead code elimination.

> Overrides are defined far away from their usage, see the example bellow.

Reviewer 3: ```suggestion
Overrides are defined far away from their usage, see the example below.
```

#### No special prop for passing overrides to children

For handling multiple levels of overrides, instead of using a special prop with all its problems (see above), `useCSS` just returns a class which can be passed to a child using `className`. If there is another `useCSS` call in the child, it takes the `className` from parent.

1. `useCSS` call in parent generates a classname, and adds mapping from the classname to a style object to a global dictionary.
2. Parent takes the class returned by the `useCSS` call and passes it to its child.
3. The child calls `useCSS` to apply its own overrides and also passes it the class from parent.
4. `useCSS` finds the style object in the dictionary, merges the original style object with the child's style overrides and returns the new classname. Instead of performing an expensive object deep merge, it just concatenates the two stringified style objects and depends on the-latest-wins CSS rule.

> ![](makeStyles-useCss.png)

Reviewer: Can we add alts to images?

Although the string concatenation is performant, it results in a lot of CSS code.

### v8 `makeStyles`

In order to improve rendering performance, v8 approach is closer to static CSS.

There are two main functions in this approach `makeStyles` and `useStyles`.
`makeStyles` is a factory which takes style object as input and returns a `useStyles` hook. `makeStyles` does not accept any theme or React context and should be called in a module scope, not in a scope of a React component.
When rendering a component, `useStyles` hook is called passing a theme to it. The `useStyles` function injects the styles into DOM (if not done yet) and maps props to classnames.

#### Multiple classes are applied to a single element

As in static CSS approach, multiple semantic classes are applied to an element.
For example `<Button primary disabled>` renders as `<button class="primary disabled">`.
To style the component, combined classes are used in CSS as well:

```css
.primary {
}
.disabled {
}
.primary.disabled {
}
```

> This is causing problems with overrides - assuming it is guaranteed that override classes are written bellow the component styles in the DOM, for most cases in this example, it is enough to have 1 class specificity for overrides to work correctly. The only case where the 1 class override does not work is for CSS properties in `.disabled.button`.

Author: The overrides story is not completely clear ATM and will need further work. Let's solve that in a separate PR.

> Consumer does not know what specificity must be used for the overrides, this can also change between library versions. Also different engineers can beak each other's styling.

Reviewer 3: `beak`?

> #### Stylesheet is generated build time

Reviewer 3: ```suggestion
#### Stylesheet is generated in build time
```

There is a build time step which generates stylesheet.
In runtime only the classes are applied based on props and state.

#### Styles precedence is set by module evaluation order

> The order of `makeStyles` calls defines the DOM insertion order. When `makeStyles` is called, it does NOT insert the CSS to DOM, but assigns a counter-based id to the associated styles. When `useStyles` returned by the `makeStyles` is called for the first time, it inserts the CSS to DOM in the order defined by the `makeStyles` call.

Reviewer 3: ```suggestion
The order of `makeStyles` calls defines the DOM insertion order. When `makeStyles` is called, it does **not** insert the CSS to DOM, but assigns a counter-based id to the associated styles. When `useStyles` returned by the `makeStyles` is called for the first time, it inserts the CSS to DOM in the order defined by the `makeStyles` call.
```

For that to work as expected, you cannot export `makeStyles` to share styles in multiple different modules - that would lead to non-deterministic insertion order. The only way to share styles is to export the style object and have a dedicated `makeStyles` calls in all the modules which reuse the style object.

The approach also currently does not support async API (`React.Suspense`) as there the modules evaluation order is non-deterministic. (Code example: https://codesandbox.io/s/kind-davinci-xvwxg)

#### Inspired by Material UI

The approach is inspired by Material UI
There is a build step which generates a stylesheet for every `makeStyles` call.

## Converged solution

Can we use v8 `makeStyles` (build time style evaluation) but let JS handle the "specificity"?

> The expensive part in v0 styling approach is the style evaluation - processing styles, expanding CSS shorthands, vendor prefixing, RTL, generating classnames. But we can do most of this build time.

Reviewer 3: ```suggestion
The expensive part in v0 styling approach is the style evaluation - processing styles, expanding CSS shorthands, vendor prefixing, RTL, generating classnames. But we can do most of this in build time.
```

### Basic algorithm

As an input, there is a style object:

```js
source = {
  color: 'red',
  margin: '0 10px 0 0',
  ':hover': {
    background: 'green',
  },
};
```

In build time, this object can be processed (CSS expand, RTL...) and atomic CSS generated for each CSS property:

```js
// ⚠️ This is proposed and simplified example output for the purpose of the spec, not the real build output ⚠️
buildOutput = {
  color: { classname: 'abcd', css: '.abcd{color:red}' },
  marginTop: { classname: 'efgh', css: '.efgh{margin-top:0}' }, // CSS expand
  marginRight: { classname: 'ijkl', css: '.ijkl{margin-right:10px}', rtlCss: '.rijkl{margin-left:10px}' }, // expand, RTL
  // ...
  ':hover+background': { classname: 'mnop', css: '.mnop:hover{background:green}' },
};
```

This object uses CSS property names as keys, builds "compound" string keys for nested objects to flatten the output. With that approach, it should be quite cheap to apply overrides. It is not necessary to read the object values, we just need to shallow apply (merge) the override object on top of the base object. No style processing is necessary:

```js

> merged = {
>   ...baseClasses,
>   ...overrideClasses,

Reviewer: ```suggestion
const merged = {
  ...baseClasses,
  ...overrideClasses,
}
```

```

After the merge, we have all the classes which need to be applied to the element being styled.

### Mapping props/state to styles

The output styles of a component change based on its props and state.
For that the `makeStyles` call accepts an array of items where each item is a `[matcher; styles]` pair:

```js
useStyles = makeStyles([

> 	[matcher1, styles1],
> 	[matcher2, styles2],

Reviewer: ```suggestion
  [matcher1, styles1],
  [matcher2, styles2],
```

>   ...

Reviewer: ```suggestion
  // etc.
```

### Passing overrides to children using `className`

To avoid the ambivalence of using `className` vs `styles` props when passing overrides to children based on their type, we will use the dictionary side effect described in `useCSS` approach above.
Parent component passes its overrides to a child as a list of classes in `className` prop. The child, takes the input classes and references the global dictionary to get back the necessary CSS attribute names to be able to perform the merge - the merge is implemented in `ax()` function as described in #16411.

#### Alternative 1: encode the CSS property name to the hash

Instead of using the global dictionary side effect, we can use part of the hash in the classname to encode the CSS property:
![](makeStyles-classname.png)

Then when merging in a child, we can merge the classes by replacing previous classnames with the same CSS property hash.
This approach makes the classnames longer and according to the test does not make the merging any faster.

> #### Alternative 2: use special `styles` prop for overrides

Reviewer 2: what if components exposed style props like `color` but those props could only accept global tokens defined in theme like `red-500` which we could easily expand to a css variable in theme

Author: @ling1726 - could you please share an example of such a prop? I've seen something similar in `Badge` design spec but I don't think we should accept color as a `prop`. One of the reasons is that is not themeable.

Reviewer 2: Yes, badge was what drove this since there are color variations available for it. Having thought through it a bit, I would say that things like `color` props could be exposed for use in selectors rather and map those selectors to all the possible color variations.

Author: Discussed offline, let's implement ~4 components where we will solve this with what we have now. Then we will revisit and discuss the pattern/changes going forward.

> This approach makes the classnames longer and according to the test does not make the merging any faster.

Reviewer 3: Is this test described somewhere, can I see the results?

Reviewer: Yup, https://github.com/microsoft/fluentui/pull/16411, it explains implementation details and why a current approach in `master` is faster. In "Performance" section, you will find numbers. I also should still have benchmark files locally, if you're interested in details.

May be we can reference this PR here?

If the global dictionary does not perform as expected, we might return back to the previous v0 approach where we use a dedicated `styles` prop to pass overrides from parent to child.

### Another libraries using similar approach

This has been inspired by [Facebook stylex talk](https://www.youtube.com/watch?v=9JZHodNR184).
Similar approach is also used in Attlasian [compiled v5](https://github.com/atlassian-labs/compiled).

## Sample Code

```jsx
// Button.tsx
const useButtonStyles = makeStyles([
  [null, { cursor: 'pointer' }], // always applied

>   [props => props.primary, tokens => ({ backgroundColor: tokens.colors.primary.background })],

Reviewer 4: **q:** 

- who injects `tokens` ?
```
[..., tokens => ({ backgroundColor: tokens.colors.primary.background })]
```

- with this API, how can one use dynamic props values within style definition?

```
// contrived example
[props => props.type === 'primary', {color: tokens.colors[props.type].background }]
```
 Is this designed like this on purpose ? So if we want different styling, we need to add another tuple to the array resolution ? (increasing codes size)

Reviewer: > Is this designed like this on purpose ? So if we want different styling, we need to add another tuple to the array resolution ? (increasing codes size)

Purpose of this design is to produce the same classes always, in this case we can make there operations build time. If there is anything conditional on right, build time optimizations are no longer possible:

```jsx
{color: 'red' }, // will produce an atomic class
{ color: tokens.colors.primary.background }, // tokens will become CSS variables, again can become an atomic class 
{ color: tokens.color[props.type].background }, // ?! 💣 
```

Reviewer: > who injects tokens ?

`tokens` or actually `theme` object is injected by `ThemeProvider` (`react-theme-provider`) or `FluentProvider` (`react-provider`). `react-make-styles` consumes required contexts and wires `makeStyles()` with proper `tokens` and RTL settings.

Reviewer 4: **another q:**
- so dynamic properties in style definition part:
   -  will be solely injected from Context ?
   -  to define dynamic styles one has to use Callback api (with aforementioned data injection)
- how will type checking work ?
   - do I need to explicitly provide generic type ? `makeStyles<Props,Context>` ? or it's gonna be defined (for context) on global provider level with ability to extend on hierarchical providers ?  

Can we add this info the the spec as well please (typescript is important part of our API's) ?

Author: > If there is anything conditional on right, build time optimizations are no longer possible

That's true, but there are cases where you need to map from a prop to a shared color - there are currently 49 shared colors, that would mean you need to implement, maintain (and always execute) 49 different matchers. Let's try to thing if there is anyway how to handle this - programatically setting a css variable for component token as inline style for example?

cc @assuncaocharles as he has already hit this wall in `Badge`.

Reviewer: > - will be solely injected from Context ?

As mentioned before, we will transform tokens usage to CSS variables for browsers that support this (this also allows to perform build time optimizations):

```jsx
const useStyles = makeStyles([
  [
    null,
    theme => ({
      backgroundColor: theme.alias.color.neutral.neutralBackground1,
      color: theme.alias.color.neutral.neutralForeground1
    })
  ]
]);
```

Will result in:

```css
.f16th3vw {
  color: var(--alias-color-neutral-neutralForeground1);
}
.fpnz0vo0 {
  background-color: var(--alias-color-neutral-neutralBackground1);
}
```

Currently, [`ThemeProvider`](https://github.com/microsoft/fluentui/tree/master/packages/react-theme-provider) injects both things: an object to React Context and CSS variables.

> - to define dynamic styles one has to use Callback api (with aforementioned data injection)

Currently, there are no dynamic styles, you can only use tokens from a theme in the same way as CSS variables:

```tsx
const useStyles = makeStyles([
  [
    null,
    theme => ({
      color: theme.alias.color.neutral.neutralForeground1,
      // effect is the same, but there are no type-checks and it work not in IE11
      color: "var(--alias-color-neutral-neutralForeground1)"
    })
  ]
]);
```

Reviewer: > how will type checking work ?
> do I need to explicitly provide generic type ? makeStyles<Props,Context> ? or it's gonna be defined (for context) on global provider level with ability to extend on hierarchical providers ?

[`react-make-styles`](https://github.com/microsoft/fluentui/tree/master/packages/react-make-styles) version has only single generic type, `Tokens` are hard-coded to be a `Theme` from `react-theme`. Types for `Selectors` in `makeStyles<Selectors>` are necessary.

Reviewer: https://github.com/microsoft/fluentui/pull/16793 has an example of converged styles, it might clarify usage. I appreciate any suggestions how we can improve current document 👍

Author: This is in the spec:

> ### Applying theme to styles
> No matter what theme is used, the component styles are always the same. The only way how to change the component appearance is through tokens which can be used in style values.
> The `styles` passed to `makeStyles([[matcher, styles]])` call can be either an object or a function. The function is called with all theme tokens.
> Those tokens are resolved to CSS variable usages in build time. `ThemeProvider` component is responsible for setting the CSS variables in DOM and changing them when theme changes. When theme is switched, only the variables are changed, all styles remain the same.

@Hotell do you think it is sufficient or is there still anything what should be added to the spec?

Reviewer 4: Overall I'd be great to add some graphics or more code examples to make it easier to grok.

From my experience when using types for implementation, it should be part of any spec to showcase the interfaces ( which again makes it more "obvious" ).

**For example (simplified example - no need to go deeper imho):**

```
// type annotation of proposed API
function makeStyles<Props>(cssDefinitions: Array<[Matchers<Props>,Styles]>): Hook
// where
type Hook = (props:Props)=>string
type Props = Record<...>
type Matchers<P> = (props:P)=>boolean
type Styles = (theme:Theme)=>CssDefinition
// where
// global tokens/theme definition provided in ThemeProvider(React Context)
type Theme = {}
``` 

If I saw that I would have not so many questions for sure :D 

Appreciate your answers though 👍

> const useButtonStyles = makeStyles([

Reviewer 3: I'll piggyback here and add that it'd be useful to the newcomers such as me to point out that we should explicitly specify the generic type here

```ts
const useButtonStyles = makeStyles<TState>([
```

This is obvious now but wasn't when I was working out my first component.

>   [props => pros.disabled, { backgroundColor: 'gray' }],

Reviewer 2: should we add an example of overriding a css variable ? i.e. design token:

```typescript
useOverrideToken = makeStyles([[
  null,
  () => ({'--token-color': 'red})
]])
```

Reviewer: I am not sure that we want to promote this approach as it's unclear how it can work in IE11 for me.

Author: The overrides story is not completely clear ATM and will need further work. Let's solve that in a separate PR.

]);

const Button = ({ className, ...props }) => {
  const buttonClasses = useButtonStyles(props); // button styles
  const mergedClasses = ax(buttonClasses, className); // merge with `className` = overrides from parent
  return <button className={mergedClasses} {...props} />;
};

// ButtonUsage.tsx
const useButtonOverrides = makeStyles([[null, { backgroundColor: 'red' }]]);

const ButtonUsage = () => {
  const buttonOverrides = useButtonOverrides({});

  return <Button classname={buttonOverrides}>I have always red background</Button>;
};
```

## Issues, open questions

### Theme-level style overrides

As described in _Applying theme to styles_, a theme can only define tokens but can never change styles. There is a concern that this is not flexible enough and there might be a valid scenario where theme-specific style overrides will be required.
How would that be implemented.
Example API:

```js
// The Button styles need an id to pull overrides
// from the theme:

> const makeStyles({ id: 'Button', { ... } });

Reviewer 4: **q:** this is some different syntax on purpose ? (no array with tuples)

Reviewer: It's not a real code, it only highlights an idea for a solution where we need to have theme specific overrides. Currently only this thing is possible:

```tsx
const useStyles = makeStyles([
  [null, { color: "black", borderWidth: "2px" }],
  [{ isDarkTheme: true }, { color: "white", borderWidth: "0" }] // should be always dark and have no border
]);
//
useStyles({ isDarkTheme: false /* some condition based on React Context */ });
```

_The main question is why not use tokens for this use case?_ Sometimes for overrides there are no proper tokens. We don't know if it's a pattern, that's why it stays as a proposal.

Reviewer 4: Can this extended example be added to the doc as well ?

Author: added to `Sample Code` section

// The theme needs to be able to define them:
const theme = {
  styles: {
    Button: { // id?
      styles: makeStyles({ ... }) // maybe?
    }
  }
}
```

### Always matching matcher

What is the best signature for a `makeStyles` call with a single style object which always matches?

```js
makeStyles([[null, { background: 'red ' }]]);

> makeStyles([{ background: 'red' }]);

Reviewer 4: **thought:** 
- `[[null, {stylesDef}]]`: explicit need to use `null` for positional args, feels rather as an anti-pattern. also this isn't very "user friendly" syntax
- `[{stylesDef}]`: need of wrapping pure object into an array feels kinda strange, on the other hand makes it more consistent (also can be explicitly typed / enforce that only first array item can contain non tuple)
- `{stylesDef}`: cleanest way, although will get messy when dynamic rules need to be applied

**Q (probably out off topic)**

Is there any particular reason why we cannot embrace "common cssInJS APIs for defining dynamic definitions", like  following ( I assume ability to process rules in compile step ? )
```
makeStyles((props:ButtonProps) => ({
   // static definitions
   color: 'red';
   '&:focus,&:hover': { ... },
   // dynamic defs
   backgroundColor: props.primary ?? tokens.colors.primary.background
}))
```

Reviewer 4: **q:** What is expect to happen if user provides multiple "apply all rules" ?

```
  [null, { cursor: 'pointer' }], // always applied
  [props => props.primary, tokens => ({ backgroundColor: tokens.colors.primary.background })]
  [null, { color: 'red' }], // always applied
```

I assume it will be merged right? do we wanna allow this from API standpoint ( if not , my previous comment regarding `[{stylesDef}]` would be great candidate for "always applied styles" )

Reviewer: > Is there any particular reason why we cannot embrace "common cssInJS APIs for defining dynamic definitions", like following ( I assume ability to process rules in compile step ? )

`backgroundColor: props.primary ?? tokens.colors.primary.background` there is no way to make this build time compatible, even if we will replace `props.primary`, for example, it will not produce the same thing.

[Linaria](https://github.com/callstack/linaria) and [Compiled](https://compiledcssinjs.com/) have similar restrictions.
